KB-12088 |DEV | BE | Enhancements in the Event enrolment API to validate the course completion check.

tarentomaheshvakkund opened the pull request:
1.  Enhancements in events enrolment API to validate if the user has completed 30 percent of the course linked to enrol to the event.
2. If the status is 2 it means the user is already completed the course.
3. If not then we will iterate through all the content completion for all the language and find the max  progress.
4. Once the max progress is fetched will compute the completion percentage and accordingly restrict the user from enrolling to the event.
4. The completion percentage is made configurable .

tarentomaheshvakkund committed:
KB-12088 |DEV | BE | Enhancements in the Event enrolment API to validate the course completion check.

1.  Enhancements in events enrolment API to validate if the user has completed 30 percent of the course linked to enrol to the event.
2. If the status is 2 it means the user is already completed the course.
3. If not then we will iterate through all the content completion for all the language and find the max  progress.
4. Once the max progress is fetched will compute the completion percentage and accordingly restrict the user from enrolling to the event.

karthik-tarento commented on course-mw/course-actors/src/main/java/org/sunbird/learner/actors/coursebatch/EventsActor.java:
```diff
@@ -377,6 +383,38 @@ private void eventEnroll(Request request) throws Throwable {
+    private void validaSamuhikCharchaEnrolment(RequestContext requestContext, String userId, String batchId, Map<String, Object> contentDetails) {
+        Integer leafNodesCount = (Integer) contentDetails.get(JsonKey.LEAF_NODE_COUNT);
+        Map<String, Object> primaryKey = new HashMap<>();
+        primaryKey.put(JsonKey.BATCH_ID, batchId);
```
this batchId is coming from which request??
Do we need batchId to check course progress? simply read userId + courseId and get all the records and identify the active one... ??

tarentomaheshvakkund replied:
BatchID was passed mistakenly so removed it  , now fetching all the records and in that filtering only the active records if it is 0 or more than 1 throwing error.

karthik-tarento commented on course-mw/course-actors/src/main/java/org/sunbird/learner/actors/coursebatch/EventsActor.java:
```diff
@@ -377,6 +383,38 @@ private void eventEnroll(Request request) throws Throwable {
+        Response response = cassandraOperation.getRecordByIdentifier(requestContext, JsonKey.KEYSPACE_SUNBIRD_COURSES, JsonKey.USER_ENROLMENTS_V2, primaryKey, null);
+        List<Map<String, Object>> userCourseEnrolmentList =
+                (List<Map<String, Object>>) response.get(JsonKey.RESPONSE);
+        if ((Integer) userCourseEnrolmentList.get(0).get("status") != 2) {
```
Need a loop here and we need to make sure there is only one active enrolment -- if there are zero or more than one then return error - which should say course completion is required in order to enrol into event.

tarentomaheshvakkund replied:
error is thrown if zero or more than one record exists.

karthik-tarento commented on course-mw/sunbird-util/sunbird-platform-core/common-util/src/main/java/org/sunbird/common/responsecode/ResponseMessage.java:
```diff
@@ -487,6 +487,7 @@ interface Message {
     String PRE_ENROLLMENT_RESOURCE_NOT_COMPLETED = "Pre-Enrollments Criterias are enabled for the course but not completed by the user.";
     String BATCH_SIZE_EXCEEDED_VALUE = "Max batch size exceeded. Only {0} more users can be enrolled to this batch.";
     String BATCH_SIZE_NOT_DEFINED_VALUE = "Batch size is not defined for this course batch.";
+    String MINIMUM_ENROLLMENT_CRITERIA_NOT_MET_VALUE = "Minimum enrollment criteria not met for this Event.";
```
Get the error message from Product team

tarentomaheshvakkund replied:
updated the error message confirming with the product.

tarentomaheshvakkund committed:
KB-12088 |DEV | BE | Enhancements in the Event enrolment API to validate the course completion check.

1. We will fetch all the enrolments records for the course passing the userid and courseid.
2.If there are zero or more than one active records then throwing error.
3.BatchId was passed wrongly so removed it.

tarentomaheshvakkund committed:
Updated the Error message for the enrollment of the events.

tarentomaheshvakkund committed:
Review comments fixed.
